### DISCUSSION TITLE
conditioning actions on serial codes

### DISCUSSION THREAD

**Discussion by preinagel** [bonsai-rx/bonsai#1545] [Q&A] [answered]
This workflow displays a hardwired stimulus whenever any serial code is received on serial port COM4, and the stimulus stays on until an up arrow key is pressed. 
![workflow](https://github.com/bonsai-rx/bonsai/assets/80356291/74d2428b-37a8-4e6c-ab92-2f247fcaf2c5)

where CreateObservable looks like this:
![observable](https://github.com/bonsai-rx/bonsai/assets/80356291/3cf6a21f-e385-4ebb-9c9a-d2440330a2d9)

Where I need help is making the action contingent on the code received. 
-How can I make the "SubscribeWhen" node subscribe only when the code is a member of a specified list of codes (e.g. only on 'APP_SoftCode1' or 'APP_SoftCode2')? 
-How can I make the stimulus parameter value also determined by the code? For example instead of the "float" which is the input to CreateObservable, can I make a node that outputs an arbitrary specified float depending on the code, like 'APP_SoftCode1' -> 0, 'APP_SoftCode2'->3.5?  (seems like ToDictionary might be what I need but don't know how to use it).

Finally I can't tell if I've set up the repeat structure right, will this just listen for codes and do the specified actions indefinitely?

thanks!

[GoToStimProtocolTrySoftcodes230901.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/12500544/GoToStimProtocolTrySoftcodes230901.bonsai.zip)

**Comment by bruno-f-cruz** [comment]
-How can I make the "SubscribeWhen" node subscribe only when the code is a member of a specified list of codes (e.g. only on 'APP_SoftCode1' or 'APP_SoftCode2')?
Use a `Condition` that checks if a string matches what you expect using `Equal`

-How can I make the stimulus parameter value also determined by the code? For example instead of the "float" which is the input to CreateObservable, can I make a node that outputs an arbitrary specified float depending on the code, like 'APP_SoftCode1' -> 0, 'APP_SoftCode2'->3.5? (seems like ToDictionary might be what I need but don't know how to use it).

Sure, you can preload it as a dictionary. I would just cut the middle man and code it straight into the message itself. You can then use the `Parse` node and extract the necessary information.

**Reply by preinagel** [reply]
OK, let me ask in a different way. I'm a confused beginner; my questions are super basic. Visual programming and reactive coding are both foreign to me. 
_Can you send me an example workflow fragment that compares the output of a SerialRead to a reference value and produces an output that can be used as an input to SubscribeWhen?_

Context in more detail if you need it:
- I have a SerialRead node that is successfully reading a byte from a SerialPort, which I know because if I connect the SerialRead node output to a SerialWrite node, it will return the value of the byte that was sent.
- Currently the SerialRead node is directly connected to a SubscribeWhen node, such that **any** byte arriving on the SerialPort triggers a visual stimulus to come on (BonVision). This works fine.  
- But what I want is for the stimulus to come on only when the received byte is a specific value, say, 5.  I tried sticking an Equal node between the SerialRead node and the SubscribeWhen node and setting the value field of the Equal node to 5, but this is evidently wrong. It throws an error "Binary operator cannot be applied to input arguments when a default operand is already specified". I am not sure if I am using Equal node incorrectly or not understanding what kind of output SerialRead is emitting or both.

When that much is working I will see how far I can get from there before asking another question.

Thanks a million for answering questions on this forum!

**Reply by bruno-f-cruz** [reply]
Sure, take a look at this tutorial, it will show you how to use conditions to filter sequences:
https://bonsai-rx.org/docs/tutorials/state-machines.html#exercise-5-specifying-conditional-task-outcomes

The idea is to use the output of the SerialReadLine and evaluate it inside the condition.

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/5c13b493-2650-44a6-83bc-f1e885096e7f)

**Reply by preinagel** [reply]
Thanks, that clarifies what you meant by "use Condition". I'll review that tutorial, but I don't think it answers the following question:

The suggested workflow doesn't crash but doesn't work. I don't know what value to put in the Equal node. What string should SerialReadLine emit if the byte value 5 is received on the serial line? If I run the visualizer on the SerialReadLine node I don't see any incoming events when I send bytes to the serial port, however.

If I use SerialRead instead of SerialReadLine I do see the incoming bytes in the visualizer of the SerialCode node (they are time stamped at the correct times and plotted to the correct byte values). Yet if I feed the output of that SerialRead node directly into Equal node (after setting up the Equal node to expect a Byte by feeing a Byte node into it), I get the error code noted above. Inserting the Condition node did not solve that problem. If I just feed a Byte node directly into the Equal node setting them to the same values, it works as expected (the SubcribeWhen event is triggered as soon as the workflow is started).    What format is the output of SerialRead node, and how do I change it into the format of 'Byte' so that Equal (expecting a byte) works?   

(Is there a general way to inspect the type of the output of a node?)

**Reply by preinagel** [reply]
hi @bruno-f-cruz   could you please post the workflow file for your response above instead of just an image?

**Reply by bruno-f-cruz** [reply]
Sure!

Additionally, regarding your previous question. Is there any reason why you are sending data as bytes instead of sending already formatted strings (at the end of the day they are all bytes, but sending strings tends to be more user-friendly for these sort of applications, additionally, the operator will automatically parse delimiter characters).

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:cv="clr-namespace:Bonsai.Vision;assembly=Bonsai.Vision"
                 xmlns:port="clr-namespace:Bonsai.IO.Ports;assembly=Bonsai.System"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="cv:CameraCapture">
          <cv:Index>0</cv:Index>
          <cv:CaptureProperties />
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="port:SerialReadLine" />
      </Expression>
      <Expression xsi:type="Equal">
        <Operand xsi:type="StringProperty">
          <Value>StringCode</Value>
        </Operand>
      </Expression>
      <Expression xsi:type="rx:Condition">
        <Workflow>
          <Nodes>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:SubscribeWhen" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="4" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="3" To="4" Label="Source2" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>

**Reply by preinagel** [reply]
I meant the visual workflow. Just wanted to be able to click on the elements and inspect their settings. 
Where would one paste such code?

**Reply by bruno-f-cruz** [reply]
Copy paste it onto the bonsai canvas

**Reply by preinagel** [reply]
To answer your question, I am sending soft codes from Bpod. To my knowledge the only soft code Bpod can send to the app com port is a byte with a value between 0 and 15.

**Reply by preinagel** [reply]
what's the "bonsai canvas"? tried pasting into main workflow window, doesn't work; tried saving as a text file with extension .bonsai and that didn't work either.

**Reply by bruno-f-cruz** [reply]
What error do you get

**Reply by preinagel** [reply, marked as answer]
when I paste in a blank workflow window, nothing happens, no error

when trying to open the file it was an XML error complaining about leading white space. 
deleted the white space and now it works. thanks.
